cabal-install/src/Distribution/Client/ScriptUtils.hs

```diff
@@ -0,0 +1,343 @@
+    ( createDirectoryIfMissingVerbose, createTempDirectory, die', handleDoesNotExist, readUTF8File, warn )
+import qualified Distribution.SPDX.License as SPDX
+import Distribution.Solver.Types.SourcePackage as SP
+    ( SourcePackage(..) )
+import Distribution.Types.BuildInfo
+    ( BuildInfo(..) )
+import Distribution.Types.CondTree
+    ( CondTree(..) )
+import Distribution.Types.Executable
+    ( Executable(..) )
+import Distribution.Types.GenericPackageDescription as GPD
+    ( GenericPackageDescription(..), emptyGenericPackageDescription )
+import Distribution.Types.PackageDescription
+    ( PackageDescription(..), emptyPackageDescription )
+import Distribution.Types.PackageName.Magic
+    ( fakePackageCabalFileName, fakePackageId )
+import Distribution.Utils.Path
+    ( unsafeMakeSymbolicPath )
+import Distribution.Verbosity
+    ( normal )
+import Language.Haskell.Extension
+    ( Language(..) )
+
+import Control.Concurrent.MVar
+    ( newEmptyMVar, putMVar, tryTakeMVar )
+import Control.Exception
+    ( bracket )
+import qualified Data.ByteString.Char8 as BS
+import Data.ByteString.Lazy ()
+import System.Directory
+    ( canonicalizePath, doesFileExist, getTemporaryDirectory, removeDirectoryRecursive )
+import System.FilePath
+    ( (</>), dropFileName, takeFileName )
+import qualified Text.Parsec as P
+
+
+-- | Get the directory where script builds are cached.
+--
+-- @CABAL_DIR\/script-builds\/@
+getScriptCacheDirectoryRoot :: IO FilePath
+getScriptCacheDirectoryRoot = do
+  cabalDir <- getCabalDir
+  return $ cabalDir </> "script-builds"
+
+-- | Get the hash of a script's absolute path)
+--
+-- Two hashes will be the same as long as the absolute paths
+-- are the same.
+getScriptHash :: FilePath -> IO String
+getScriptHash script = showHashValue . hashValue . fromString <$> canonicalizePath script
+
+-- | Get the directory for caching a script build.
+--
+-- The only identity of a script is it's absolute path, so append the
+-- hashed path to @CABAL_DIR\/script-builds\/@ to get the cache directory.
+getScriptCacheDirectory :: FilePath -> IO FilePath
+getScriptCacheDirectory script = (</>) <$> getScriptCacheDirectoryRoot <*> getScriptHash script
+
+-- | Get the directory for caching a script build and ensure it exists.
+--
+-- The only identity of a script is it's absolute path, so append the
+-- hashed path to @CABAL_DIR\/script-builds\/@ to get the cache directory.
+ensureScriptCacheDirectory :: Verbosity -> FilePath -> IO FilePath
+ensureScriptCacheDirectory verbosity script = do
+  cacheDir <- getScriptCacheDirectory script
+  createDirectoryIfMissingVerbose verbosity True cacheDir
+  return cacheDir
+
+-- | What your command should do when no targets are found.
+data AcceptNoTargets
+  = RejectNoTargets -- ^ die on 'TargetSelectorNoTargetsInProject'
+  | AcceptNoTargets -- ^ return a default 'TargetSelector'
+  deriving (Eq, Show)
+
+-- | Information about the context in which we found the 'TargetSelector's.
+data TargetContext
```

Comment:
It seems to me that this, `AcceptNoTargets` and `withContextsAndSelectors` should live in a more generic module, maybe one with other target related definitions

Reply:
`withContextAndSelectors` I think makes since in a `ScirptUtils` as it's specifically a utility function for helping commands deal with script targets. I guess `withContextAndSelectors, AcceptNoTargets(..), TargetContext(..),` could be in their own module for `SciptTargetUtils` (open to name suggestions), but I'm nor sure what the benefit is.

I'm not entirely happy with the name `ScirptUtils`. Maybe a solution is to find a better name for the module as a whole instead of splitting it?

Reply:
Hmm the signature of the function, its name and types are generic and is being used for run, build and repl, as a higher order function.
It is a little bit surprising see it living in script utils only looking to the signature.
But its implementation is mainly script related, as it prapares context and selectors for the script case.

Idk, maybe changing the name to `withScriptContextAndSelectors` could help? As you consider.

Then if you look at the implementation, it is handling three cases:

Reply:
I don't love the name `TargetContext` because Context is so heavily overloaded, but I didn't know what else to call it.  I'm not sure I like `withScriptContextAndSelectors` because it implies it's specific to when there is a script rather than optionally accepting scripts.